sys/net/gnrc/routing/rpl/mrhof.c

```diff
@@ -0,0 +1,288 @@
+    }
+
+    if (p2_acceptable == false) {
+        /* p1 is acceptable and p2 not */
+        return -1;
+    }
+
+    return 0;
+}
+
+/* Compare freshness of two parents */
+static int _mrhof_cmp_fresh(netif_t *netif,
+                            netstats_nb_t *p1_stats,
+                            netstats_nb_t *p2_stats)
+{
+    bool p1_isfresh = _mrhof_isfresh(netif, p1_stats);
+    bool p2_isfresh = _mrhof_isfresh(netif, p2_stats);
+
+    if (p1_isfresh == p2_isfresh) {
+        return 0;
+    }
+
+    /* One of the two parents is not acceptable */
+    if (p1_isfresh == false) {
+        /* p2 is acceptable and p1 not */
+        return 1;
+    }
+
+    if (p2_isfresh == false) {
+        /* p1 is acceptable and p2 not */
+        return -1;
+    }
+
+    return 0;
+}
+
+static void reset(gnrc_rpl_dodag_t *dodag)
+{
+    /* Nothing to do in MRHOF */
+    (void) dodag;
+}
+
+/**
+ * Calculate rank additive based on the rank of the parent and the etx to the parent
+ * computed via MAX(pref_parent->rank, 1 + floor(MAX(parents->rank)/MinHopRankIncrease),
+ */
+static uint16_t calc_rank(gnrc_rpl_dodag_t *dodag, uint16_t base_rank)
+{
+    (void) base_rank;
+    DEBUG("MRHOF: Calculating rank\n");
+
+    /* TODO: consider a parent set of more than 1 parent */
+    if (dodag == NULL) {
+        DEBUG("MRHOF: No dodag, assuming max rank\n");
+        return GNRC_RPL_INFINITE_RANK;
+    }
+    else {
+        gnrc_rpl_parent_t *elt = NULL;
+        netstats_nb_t elt_stats;
+        uint8_t cnt = 0;
+        uint8_t max_dagrank = 0;
+        uint16_t cost_rank, minhoprankincr = dodag->instance->min_hop_rank_inc;
+
+        /* Determine the path cost through the preferred parent */
+        if (!_mrhof_get_stats(dodag->parents, &elt_stats)) {
+            DEBUG("MRHOF: No stats for parent, assuming max rank\n");
+            return GNRC_RPL_INFINITE_RANK;
+        }
+        cost_rank = _mrhof_get_path_cost(dodag->parents, &elt_stats);
+
+        /* Determine the largest dagrank in the parent set */
+        LL_FOREACH(dodag->parents, elt) {
+            if (cnt >= MRHOF_PARENT_SET_SIZE) {
+                break;
+            }
+            _mrhof_get_stats(elt, &elt_stats);
+            /* Only include parent in the parent set if the statistics are acceptable
+             * and the path cost is not significantly worse than the current preferred parent */
+            if (_mrhof_is_acceptable(elt, &elt_stats) && \
+                (_mrhof_get_path_cost(elt, &elt_stats) <= cost_rank + MRHOF_PARENT_SWITCH_THRESHOLD)) {
+                uint8_t new_dagrank = DAGRANK(elt->rank, minhoprankincr);
+                if (max_dagrank < new_dagrank) {
+                    max_dagrank = new_dagrank;
+                }
+            }
+            else {
+                /* Break when parents are no longer acceptable */
+                break;
+            }
+            cnt++;
+        }
+        uint16_t monotonic_rank = minhoprankincr * (max_dagrank + 1);
```

Comment:
Is this maybe 3? I can't find anything about "monotonic rank" in RFC 6719. If so, I would prefer the wording from the RFC.

Reply:
This is part of 2, the loop above gets the max advertised dagrank among the set of parents, this calculates the rank to advertise based on the incremented dagrank.